Fix docs regarding deduplication periods [KVL-1194] (#11738)
digital-asset/daml: docs/source/app-dev/services.rst

- Applications can provide a :ref:`deduplication duration <com.daml.ledger.api.v1.Commands.deduplication_duration>` for each command. If this parameter is not set, the default maximum deduplication period is used.
- A command submission is considered a duplicate submission if the ledger API server is aware of another command within the deduplication period and with the same :ref:`change ID <change-id>`.

> - Duplicate command submissions will generate a rejection until either the effective deduplication period of the original command has passed or the original submission was rejected (i.e. the command failed and resulted in a rejected transaction), whichever comes first.

Reviewer: The backwards looking will land in the 2.0 release. The PR is ready, but we need a follow up PR and decided not to rush for this release

> - Duplicate command submissions will generate a rejection until either the effective deduplication period of the original command has passed or the original submission was rejected (i.e. the command failed and resulted in a rejected transaction), whichever comes first.

Author: Good point, applied.

For details on how to use command deduplication, see the :ref:`Application Architecture Guide <command-deduplication>`.

.. note:: Ledgers with command deduplication across participants

>   - The ledger server may extend the deduplication period specified in the request arbitrarily, up until the maximum deduplication duration specified in the :ref:`ledger configuration <ledger-configuration-service>`.

Reviewer 2: ```suggestion
  - The ledger may extend the deduplication period specified in the request arbitrarily, up until the maximum deduplication duration specified in the :ref:`ledger configuration <ledger-configuration-service>`.
```

Reviewer: Do we want to capture that it also accounts for time skews? Currently we basically account for `min_record_time` by adding the `min_skew` to calculate the deduplicate until

>     The deduplication period chosen by the ledger server is the *effective deduplication period*.

Reviewer 2: ```suggestion
    The deduplication period chosen by the ledger is the *effective deduplication period*.
```

  - Regardless, the deduplication period specified in the request is always checked against the configured maximum deduplication duration.

>   - A command submission is considered a duplicate submission if the ledger server is aware of another command within the *effective* deduplication period and with the same :ref:`change ID <change-id>`.

Reviewer 2: ```suggestion
  - A command submission is considered a duplicate submission if the ledger is aware of another command within the *effective* deduplication period and with the same :ref:`change ID <change-id>`.
```

>   - Command deduplication is *guaranteed* to work even if commands are submitted to different participants.

Reviewer 2: Is this true for VMBC and Canton? Maybe we can just mention both in this sentence.

Author: Sounds good. This is certainly true for VMBC, @andreaslochbihler-da can comment on Canton.

Author: @gerolf-da Opted for a slightly different presentation (and left out Canton till I get a confirmation) -- let me know what you think.

Reviewer 3: Command deduplication is formally defined via the completion event stream. Since Canton produces a completion event only on the participant that made the submission, Canton doesn't deduplicate commands across participants.

Author: OK, removed it.

For details on how to use command deduplication, see the :ref:`Application Architecture Guide <command-deduplication>`.

> .. note:: Ledgers with command deduplication across participants

Reviewer 3: ```suggestion
.. note::
```

>   - The ledger may extend the deduplication period specified in the request arbitrarily, up until the maximum deduplication duration specified in the :ref:`ledger configuration <ledger-configuration-service>`.

Reviewer 3: ```suggestion
  - The ledger may extend the deduplication period specified in the request arbitrarily, even beyond the maximum deduplication duration specified in the :ref:`ledger configuration <ledger-configuration-service>`.
```

Author: Are you suggesting 'even beyond' because of accounting for time skew or for a different reason?

Reviewer 3: Time skews are one thing where it can already happen for kv-based ledgers.
The other thing is that Canton has to report a deduplication offset rather than a duration; but if the submission had specified a duration, then there may not be a suitable completion offset around the time where the deduplication duration starts. So if the deduplication duration already is max dedup time, we cannot really return an offset that was published within the max dedup duration without incurring the risk that we're shortening the effective deduplication duration. So instead, Canton doesn't even bother to find a completion offset close the the beginning of the deduplication duration. Instead, we return the first known offset on the ledger or the offset after the last completion for the change ID. So for an unpruned ledger and a fresh change ID, the effective deduplication period is unbounded. Given that longer durations provide a stronger statement to the application, the application should actually be happy about this.

Reviewer 3: If there is already a completion for the same change ID, then Canton typically does not extend the deduplication period before the completion if the earlier completion wouldn't normally fall into the deduplication period. But as always, with clock drift, this may appear to the application as if some extension happened.

  - Command deduplication is *guaranteed* to work even if commands are submitted to different participants.

  The following ledger integrations support command deduplication across participants:
    - `Daml Driver for VMware Blockchain <https://www.digitalasset.com/daml-for-vmware-blockchain/>`__

>     - :ref:`Daml Sandbox <sandbox-manual>`

Reviewer 2: There really is only one single participant. So while the statement technically is true, it's not very helpful.

Author: You mean for Daml Sandbox, right? I've removed that and put this note above where we mention 'command deduplication across participants'.
